proposals/0000-hyperdb.md

```diff
@@ -0,0 +1,581 @@
+
+The HyperDB specification (this document) is complicated by the inclusion of
+new protobuf fields to support "multi-writer" features which are not described
+here. The motivation to include these fields now to make only a single
+backwards-incompatible schema change, and to make a second software-only change
+in the future to enable support for these features. Schema and data format
+changes are considered significantly more "expensive" for the community and
+software ecosystem compared to software-only changes. Attempts have been made
+in this specification to indicate the safe "single-writer-only" values to use
+for these fields.
+
+
+# Dat migration logistics
+[migration]: #migration
+
+HyperDB is not backwards compatible with the existing hyperdrive metadata,
+meaning dat clients may need to support both versions during a transition
+period. This applies both to archives saved to disk (eg, in SLEEP) and to
+archives received and published to peers over the network.
+
+No changes to the Dat network wire protocol itself are necessary, only changes
+to content passed over the protocol. The Dat `content` feed, containing raw
+file data, is not impacted by HyperDB, only the contents of the `metadata`
+feed.
+
+Upgrading a Dat (hyperdrive) archive to HyperDB will necessitate creating a new
+feed from scratch, meaning new public/private key pairs, and that public key
+URL links will need to change.
+
+Further logistical details are left to the forthcoming Multi-Writer DEP.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+How are hash collisions handled? Eg, two keys which have the same `path` hash
+array. How unlikely is this situation? Eg, how many keys in a database before
+the birthday paradox results in a realistic chance of a collision? How should
+implementations behave if they detect a collision (matching path but not
+matching key)?
+
+Need to think through deletion process with respect to listing a path prefix;
+will previously deleted nodes be occulded, or potentially show up in list
+results?
+
+Are the "deletion" semantics here correct, in that deletion Nodes persist as
+"tombstones", or should deleted keys be omitted from future `trie` fields to
+remove their presence?
+
+Should all-zeros really be used for path hashing, or should we use the
+hypercore feed public key? It certainly makes implementation and documentation
+simpler to use all-zeros, and potentially makes it easier to copy or migrate
+HyperDB content between hypercore feeds. Referencing the feed public key breaks
+abstraction layers and the separation of concerns.
+
+There are implied "reasonable" limits on the size (in bytes) of both keys and
+values, but they are not formally specified. Protobuf messages have a hard
+specified limit of 2 GByte (due to 32-bit signed arthimetic), and most
+implementations impose a (configurable) 64 MByte limit. Should this DEP impose
+specific limits on key and value sizes? Would be good to decide before Draft
+status.
```

Comment:
since dat does rabin chunking, the 2GB limit seems fine ~ large files will always be divided into chunks smaller than that.

will the chunks always be below 64MB though? @mafintosh

Reply:
@dcposch note this actually not for the files but just for the file *metadata*. in our case that is just a fs.Stat object with an additional numeric pointer to another feed where the actual file content is stored. This gives basically as large files as you'd want

Reply:
Also, FWIW, dat (via hyperdrive) does not actually use Rabin chunking at the moment, just fixed size chunks. The protocol is agnostic to chunking, so this can evolve in the future with little disruption if/when content-aware chunking is more performent.

Reply:
To remind, dat (via hyperdrive) is only one possible application on top of hyperdrive; while hyperdrive should be pretty comfortable even with a couple KByte metadata size limit, other applications might want to store arbitrary data in the hyperdb values; setting expectations around this could save frustration or growing pains later.

Reply:
@bnewbold agreed. imo you should always only store "small" values directly in the hyperdb, say < 1mb. anything you should put in a "content feed" and then link it from the value.

Reply:
I think for Draft we can mention the inherent limits and then mention:

"It is STRONGLY recommended to keep both keys and values to less than 2 megabytes (individually and combined). If larger value sizes are desired, a separate content can be pointed to (via `contentFeed`), and the value in this feed can be metadata pointing to data (eg, contiguous blocks of entries) in that feed."